README.md
spl-meter
=========

SPL meter for Linux

credits to http://www.youtube.com/watch?v=KgapjPoEeJI


debian/changelog
wb-mqtt-spl-meter (1.1.1) stable; urgency=medium

  * use newer libwbmqtt

 -- Evgeny Boger <boger@contactless.ru>  Thu, 22 Dec 2016 19:41:00 +0300

wb-mqtt-spl-meter (1.1) stable; urgency=medium

  * JSON config, C++ rewrite, etc

 -- Alexey Ignatov <lexszero@gmail.com>  Thu, 08 Oct 2015 01:40:49 +0000

wb-mqtt-spl-meter (1.0) stable; urgency=medium

  * Initial release.

 -- Alexey Ignatov <lexszero@gmail.com>  Thu, 01 Oct 2015 12:54:19 +0000


pcm_min_dB.c
#include<alsa/asoundlib.h>
#include<math.h>

static char *device = "default";	/* device name */

/*
   Each sample is a 16-bit signed integer, which means
   the value can vary from -32768 to 32767.

   Size of the sample buffer = 8*1024 = 8192, which takes
   8*1024*2 = 16384 bytes.

   Sampling frequency = 48000Hz, in other words, sound
   card will record the sample value every 1/48000 second.
*/
short buffer[8*1024];

/*
   The unit of sizeof() function is byte, so we have to
   shift 1 bit to the right to calculate the real size of buffer.
*/
int buffer_size = sizeof(buffer)>>1;

/*
   Function for calculating the Root Mean Square of sample buffer.
   RMS can calculate an average amplitude of buffer.
*/
double rms(short *buffer)
{
	int i;
	long int square_sum = 0.0;
	for(i=0; i<buffer_size; i++)
		square_sum += (buffer[i] * buffer[i]);

	double result = sqrt(square_sum/buffer_size);
	return result;
}

int main(void)
{
	int err;
	unsigned int i;

	snd_pcm_t *handle_capture;	/* handle of capture */

	snd_pcm_sframes_t frames;

	// Open handle of capture
	if((err=snd_pcm_open(&handle_capture, device, SND_PCM_STREAM_CAPTURE, 0)) < 0) {
		printf("Capture open error: %s\n", snd_strerror(err));
		exit(EXIT_FAILURE);
	}

	if((err = snd_pcm_set_params(handle_capture,
				     SND_PCM_FORMAT_S16_LE,
				     SND_PCM_ACCESS_RW_INTERLEAVED,
				     1,
				     48000,
				     1,
				     500000)) < 0) {	/* 0.5s */
		printf("Capture open error: %s\n", snd_strerror(err));
		exit(EXIT_FAILURE);
	}
	printf("             ");
	fflush(stdout);

	/*
	 * Formula of dB is 20log((Sound Pressure)/P0)
	   Assume that (Sound Pressure/P0) = k * sample value (Linear!),
	   and by experiment, we found that k = 0.45255.
	*/
	double k = 0.45255;
	double Pvalue = 0;
	int dB = 0;
	int peak = 0;
	i=0;

	// Capture
	while(i<50) {
		frames = snd_pcm_readi(handle_capture, buffer, buffer_size);
		if(frames < 0)
			frames = snd_pcm_recover(handle_capture, frames, 0);
		if(frames < 0) {
			printf("snd_pcm_readi failed: %s\n", snd_strerror(err));
		}
		if(frames > 0 && frames < (long)buffer_size)
			printf("Short read (expected %li, wrote %li)\n", (long)buffer_size, frames);

		Pvalue = rms(buffer) * k;

		dB = (int)20*log10(Pvalue);
		if(dB > peak)
			peak = dB;
		int j;
		for(j=0; j<20; j++)
			printf("\b");
		fflush(stdout);
		printf("dB=%4d,Peak=%4d", dB, peak);
		fflush(stdout);
	}
	printf("\n");
	snd_pcm_close(handle_capture);
	return 0;
}


wb-mqtt-spl-meter.cpp
#include <iostream>
#include <stdexcept>
#include <fstream>
#include <cstring>
#include <utility>
#include <memory>
#include <getopt.h>
#include <math.h>

#include "jsoncpp/json/json.h"
#include <mosquittopp.h>
#include <wbmqtt/utils.h>
#include <wbmqtt/mqtt_wrapper.h>

#include <alsa/asoundlib.h>

using namespace std;

class THandlerConfig
{
	public:
		THandlerConfig(const std::string fname);

		string DeviceName;
		string AlsaDevice;
		int SampleRate;
		int Period;
		double K;
};

THandlerConfig::THandlerConfig(const string fname)
{
	// Let's parse it
	Json::Value root;
	Json::Reader reader;

	if (fname.empty()) {
		throw invalid_argument("Please specify config file with -c option");
	}

	ifstream myfile (fname);

	bool parsedSuccess = reader.parse(myfile,
			root,
			false);
	if (not parsedSuccess) {
		throw runtime_error("Failed to parse JSON\n" + reader.getFormattedErrorMessages());
	}

	DeviceName = root["device_name"].asString();
	AlsaDevice = root.get("alsa_device", Json::Value("default")).asString();
	SampleRate = root.get("sample_rate", Json::Value(16000)).asInt();
	Period = root.get("period", Json::Value(500)).asInt();
	K = root.get("k", Json::Value(0.45255)).asDouble();
}

class TMQTTSplHandler : public TMQTTWrapper
{
	public:
		TMQTTSplHandler(const TMQTTSplHandler::TConfig& mqtt_config, const THandlerConfig& config);
		~TMQTTSplHandler();
		
		void OnConnect(int rc);
		void OnMessage(const struct mosquitto_message *) {};
		void OnSubscribe(int, int, const int *) {};

		void PublishSPL(const int value);

	private:
		THandlerConfig Config;
};

TMQTTSplHandler::TMQTTSplHandler(const TMQTTSplHandler::TConfig& mqtt_config, const THandlerConfig& config)
	: TMQTTWrapper(mqtt_config)
	, Config(config)
{
	Connect();
}

TMQTTSplHandler::~TMQTTSplHandler() {}

void TMQTTSplHandler::OnConnect(int rc)
{
	printf("Connected with code %d.\n", rc);
	if(rc != 0)
		return;

	/* Only attempt to Subscribe on a successful connect. */
	string prefix = string("/devices/") + MQTTConfig.Id + "/";

	// Meta
	Publish(NULL, prefix + "meta/name", Config.DeviceName, 0, true);

	string control_prefix = prefix + "controls/dB";
	Publish(NULL, control_prefix + "/meta/type", "dB", 0, true);
	Publish(NULL, control_prefix + "/meta/readonly", "1", 0, true);
}

void TMQTTSplHandler::PublishSPL(int dB)
{
	Publish(NULL, string("/devices/") + MQTTConfig.Id + "/controls/dB", to_string(dB), 0, true);
}



int main(int argc, char *argv[])
{
	int ret, c;
	snd_pcm_t *capture;

	TMQTTSplHandler::TConfig mqtt_config;
	mqtt_config.Host = "localhost";
	mqtt_config.Port = 1883;
	mqtt_config.Id = "wb-spl-meter";

	string config_fname;
	
	while ( (c = getopt(argc, argv, "c:h:p:")) != -1) {
		//~ int this_option_optind = optind ? optind : 1;
		switch (c) {
		case 'c':
			config_fname = optarg;
			break;
		case 'p':
			mqtt_config.Port = stoi(optarg);
			break;
		case 'h':
			mqtt_config.Host = optarg;
			break;
		case '?':
			break;
		default:
			fprintf(stderr, "?? Getopt returned character code 0%o ??\n", c);
		}
	}

	try {
		mosqpp::lib_init();
		THandlerConfig config (config_fname);

		std::shared_ptr<TMQTTSplHandler> mqtt_handler(new TMQTTSplHandler(mqtt_config, config));
	    mqtt_handler->Init();

	    ret = mqtt_handler->loop_start();
		if (ret != 0)
			throw runtime_error("Couldn't start mosquitto_loop_start: " + to_string(ret));

		snd_pcm_sframes_t frames;
		char tmp[16];

		// Open handle of capture
		ret = snd_pcm_open(&capture, config.AlsaDevice.c_str(), SND_PCM_STREAM_CAPTURE, 0);
		if (ret < 0)
			throw runtime_error("PCM open failed: " + string(snd_strerror(ret)));
	
		ret = snd_pcm_set_params(capture,
						SND_PCM_FORMAT_S16_LE,
						SND_PCM_ACCESS_RW_INTERLEAVED,
						1,
						config.SampleRate,
						1,
						config.Period*1000);
		if (ret < 0)
			throw runtime_error("PCM parameters setting failed: " + string(snd_strerror(ret)));
		
		/*
		 * Formula of dB is 20log((Sound Pressure)/P0)
		   Assume that (Sound Pressure/P0) = k * sample value (Linear!),
		   and by experiment, we found that k = 0.45255.
		*/
		double Pvalue = 0;
		int dB = 0;
		int64_t square_sum = 0;
	
		unsigned int i;
	
		size_t buf_size = config.SampleRate/1000*config.Period;
		int16_t *buffer = new int16_t[buf_size];
		// Capture
		while (1) {
			frames = snd_pcm_readi(capture, buffer, buf_size);
			if (frames < 0) {
				frames = snd_pcm_recover(capture, frames, 0);
			}
			if (frames < 0) {
				cerr << "snd_pcm_readi failed" << endl;
				continue;
			}
			if (frames > 0 && frames < (long)buf_size) {
				cerr << "Warning: short read (expected " << buf_size << ", got " << frames << ")" << endl;
				continue;
			}
	
			square_sum = 0;
			for(i = 0; i < buf_size; i++) {
				square_sum += (buffer[i] * buffer[i]);
			}
			Pvalue = sqrt(square_sum / buf_size) * config.K;
	
			dB = (int)20*log10(Pvalue);
	
			snprintf(tmp, 16, "%d", dB);
			mqtt_handler->PublishSPL(dB);
		}
		snd_pcm_close(capture);
		mosqpp::lib_cleanup();
	}
	catch (const exception& ex) {
		cerr << "Error: " << ex.what() << endl;
		return EXIT_FAILURE;
	}

	return 0;
}


wb-mqtt-spl-meter.c
#include <stdlib.h>
#include <stdio.h>
#include <stdint.h>
#include <getopt.h>
#include <math.h>
#include <alsa/asoundlib.h>
#include <mosquitto.h>

#define MQTT_PREFIX "/devices/spl-meter"

static char *host = "127.0.0.1";
static int port = 1883;

static char *device = "default";
static int sample_rate = 16000;
static int period = 500;
static double k = 0.45255;

static bool debug = false;

static snd_pcm_t *capture;
struct mosquitto *mosq;

static int16_t *buffer;
static size_t buf_size;

static int parse_options(int argc, char *argv[])
{
	int ret;
	char *endptr;

	while ((ret = getopt(argc, argv, "h:p:d:r:t:D")) != -1) {
		switch (ret) {
			case 'h':
				host = optarg;
				break;

			case 'p':
				port = strtol(optarg, &endptr, 10);
				if (*endptr != 0) {
					fprintf(stderr, "'%s' is not a valid port number\n", optarg);
					return 1;
				}
				break;

			case 'd':
				device = optarg;
				break;

			case 'r':
				sample_rate = strtol(optarg, &endptr, 10);
				if (*endptr != 0) {
					fprintf(stderr, "'%s' is not a valid sample rate\n", optarg);
					return 1;
				}
				break;

			case 't':
				period = strtol(optarg, &endptr, 10);
				if (*endptr != 0) {
					fprintf(stderr, "'%s' is not a valid RMS period\n", optarg);
					return 1;
				}
				break;

			case 'k':
				k = strtod(optarg, &endptr);
				if (*endptr != 0) {
					fprintf(stderr, "'%s' is not a valid float value\n", optarg);
					return 1;
				}
				break;

			case 'D':
				debug = true;
				break;

			default:
				fprintf(stderr, "Unknown option: -%c\n", ret);
				return 1;
		}
	}

	return 0;
}

static int init_capture(void)
{
	int ret;
	// Open handle of capture
	if ((ret = snd_pcm_open(&capture, device, SND_PCM_STREAM_CAPTURE, 0)) < 0) {
		fprintf(stderr, "Error: PCM open failed: %s\n", snd_strerror(ret));
		return 1;
	}

	if ((ret = snd_pcm_set_params(capture,
					SND_PCM_FORMAT_S16_LE,
					SND_PCM_ACCESS_RW_INTERLEAVED,
					1,
					sample_rate,
					1,
					period*1000)) < 0) {
		fprintf(stderr, "Error: PCM parameters setting failed: %s\n", snd_strerror(ret));
		return 1;
	}

	return 0;
}

static int init_mosquitto(void)
{
	int ret;
	char *str;

	mosquitto_lib_init();

	mosq = mosquitto_new(NULL, true, NULL);
	if (!mosq) {
		switch (errno) {
			case ENOMEM:
				fprintf(stderr, "Error: Out of memory.\n");
				break;

			case EINVAL:
				fprintf(stderr, "Error: Invalid id and/or clean_session.\n");
				break;
		}

		mosquitto_lib_cleanup();
		return 1;
	}

	ret = mosquitto_connect(mosq, host, port, 5);
	if (ret != 0) {
		fprintf(stderr, "Error: Can't connect to MQTT broker\n");
		return ret;
	}

	str = "dB";
	mosquitto_publish(mosq, NULL, MQTT_PREFIX "/controls/dB/meta/type", strlen(str), str, 2, true);

	str = "Sound level meter";
	mosquitto_publish(mosq, NULL, MQTT_PREFIX "/meta/name", strlen(str), str, 2, true);

	mosquitto_loop_start(mosq);

	return 0;
}

int main_loop(void)
{
	snd_pcm_sframes_t frames;
	char tmp[16];

	/*
	 * Formula of dB is 20log((Sound Pressure)/P0)
	   Assume that (Sound Pressure/P0) = k * sample value (Linear!),
	   and by experiment, we found that k = 0.45255.
	*/
	double Pvalue = 0;
	int dB = 0;
	int64_t square_sum = 0;

	unsigned int i;

	// Capture
	while (1) {
		frames = snd_pcm_readi(capture, buffer, buf_size);
		if (frames < 0) {
			frames = snd_pcm_recover(capture, frames, 0);
		}
		if (frames < 0) {
			fprintf(stderr, "snd_pcm_readi failed\n");
			continue;
		}
		if (frames > 0 && frames < (long)buf_size) {
			fprintf(stderr, "Warning: short read (expected %li, got %li)\n", (long)buf_size, frames);
			continue;
		}

		square_sum = 0;
		for(i = 0; i < buf_size; i++) {
			square_sum += (buffer[i] * buffer[i]);
		}
		Pvalue = sqrt(square_sum / buf_size) * k;

		dB = (int)20*log10(Pvalue);

		snprintf(tmp, 16, "%d", dB);
		mosquitto_publish(mosq, NULL, MQTT_PREFIX "/controls/dB", strlen(tmp), tmp, 2, true);
		
		if (debug) {
			printf("%d dB\n", dB);
		}
	}
}

int main(int argc, char *argv[])
{
	int ret;

	ret = parse_options(argc, argv);
	if (ret != 0) {
		fprintf(stderr, "Usage: %s [-h host] [-p port] [-d device] [-r rate] [-t period]\n", argv[0]);
		fprintf(stderr, "    -h host        MQTT broker host [127.0.0.1]\n");
		fprintf(stderr, "    -p port        MQTT broker port [18883]\n");
		fprintf(stderr, "    -d device      ALSA capture device [default]\n");
		fprintf(stderr, "    -r rate        Sample rate in Hz [16000]\n");
		fprintf(stderr, "    -t period      RMS integrating period in milliseconds [500]\n");
		fprintf(stderr, "    -k double      magic multiplier [0.45255]\n");
		exit(EXIT_FAILURE);
	}

	buf_size = sample_rate/1000*period;
	buffer = calloc(buf_size, sizeof(buffer[0]));
	if (!buffer) {
		fprintf(stderr, "Can't allocate memory for buffer\n");
		exit(EXIT_FAILURE);
	}

	if (init_capture() != 0) {
		exit(EXIT_FAILURE);
	}

	if (init_mosquitto() != 0) {
		exit(EXIT_FAILURE);
	}

	printf("%s:%d, %s %dHz, %dms; buf_size=%d\n",
			host, port, device, sample_rate, period, buf_size);
	ret = main_loop();

	mosquitto_loop_stop(mosq, false);
	mosquitto_destroy(mosq);
	mosquitto_lib_cleanup();

	snd_pcm_close(capture);

	return ret;
}
